src/sync/pool.go

```diff
@@ -1,19 +1,25 @@
 package sync
 
-// Pool is a very simple implementation of sync.Pool. It does not actually
-// implement a pool.
+// Pool is a very simple implementation of sync.Pool.
 type Pool struct {
-	New func() interface{}
+	New   func() interface{}
+	items []interface{}
 }
 
-// Get returns the value of calling Pool.New().
+// Get returns an item in the pool, or the value of calling Pool.New() if there are no items.
 func (p *Pool) Get() interface{} {
+	if len(p.items) > 0 {
+		x := p.items[len(p.items)-1]
+		p.items = p.items[:len(p.items)-1]
+		return x
+	}
 	if p.New == nil {
 		return nil
 	}
 	return p.New()
 }
 
-// Put drops the value put into the pool.
+// Put adds a value back into the pool.
 func (p *Pool) Put(x interface{}) {
+	p.items = append(p.items, x)
```

Comment:
The upstream `sync.Pool` is cleared every time the garbage collector runs.  This prevents memory leaks from happening.  This implementation doesn't have the safe-guard.  Maybe we could limit the number of elements in the pool to some small constant or somehow expire items (ad their points) to prevent unbounded memory growth?

Reply:
Yeah did think about this. FWIU, the main motivation for the upstream clearing is not to prevent unbounded memory growth, it is the job of the code to Get and Put in pairs to prevent this, but to allow periods of high traffic to have a larger pool but to not keep the larger pool during low traffic. This is a nice to have optimization but I don't know if it's critical.

I can't think of an appropriate constant that wouldn't be too small or too large, and trying an approach of frequency tracking I think could become quite complicated. Any suggestion on an algorithm that wouldn't be too much?

Reply:
I think we can leave this optimization out for now. When in doubt, I prefer the simpler solution (where simpler usually but not always means less code). We can always add this optimization when needed.